File: src/main/java/com/footverse/order/dto/CouponResponse.java
package com.footverse.order.dto;

import java.math.BigDecimal;
import java.time.LocalDateTime;

import com.footverse.order.entity.DiscountType;

/**
 * A coupon returned to admins for management (dto-spec §14).
 *
 * @param id                the coupon id
 * @param code              the unique coupon code
 * @param name              the coupon name
 * @param description       the description, if any
 * @param discountType      the discount strategy ({@code PERCENT} / {@code FIXED})
 * @param discountValue     the discount value
 * @param minOrderAmount    the minimum order subtotal required
 * @param maxDiscountAmount the cap for percentage discounts, if any
 * @param startAt           the validity start
 * @param endAt             the validity end
 * @param usageLimit        the global usage cap ({@code null} = unlimited)
 * @param usedCount         the current usage count
 * @param enabled           whether the coupon is enabled
 */
public record CouponResponse(
        Long id,
        String code,
        String name,
        String description,
        DiscountType discountType,
        BigDecimal discountValue,
        BigDecimal minOrderAmount,
        BigDecimal maxDiscountAmount,
        LocalDateTime startAt,
        LocalDateTime endAt,
        Integer usageLimit,
        int usedCount,
        boolean enabled) {
}

File: admin/src/features/coupon/models/coupon-response.ts
import type { DiscountType } from './discount-type';

/**
 * A coupon returned for admin management (dto-spec §14). Mirrors the frozen
 * DTO field-for-field.
 *
 * `maxDiscountAmount` and `usageLimit` are optional (`BigDecimal`/`Integer`)
 * on the backend record with no `@JsonInclude(NON_NULL)`, so Jackson's
 * default inclusion serializes an unset value as an explicit JSON `null`
 * rather than omitting the key — the response can carry `null`, not only
 * `undefined`.
 */
export interface CouponResponse {
  readonly id: number;
  readonly code: string;
  readonly name: string;
  readonly description?: string;
  readonly discountType: DiscountType;
  readonly discountValue: number;
  readonly minOrderAmount: number;
  readonly maxDiscountAmount?: number | null;
  readonly startAt: string;
  readonly endAt: string;
  readonly usageLimit?: number | null;
  readonly usedCount: number;
  readonly enabled: boolean;
}
